File: problems/Minimum Add to Make Parentheses Valid/minAddToMakeValid.cs
public class Solution {
    public int MinAddToMakeValid(string S) {
        var ans = 0;
        var bal = 0;

        foreach (var item in S) {
            bal += item == '(' ? 1 : -1;
            // It is guaranteed bal >= -1
            if (bal == -1) {
                ans++;
                bal++;
            }
        }

        return ans + bal;
    }
}

File: problems/Minimum Add to Make Parentheses Valid/minAddToMakeValid.java
class Solution {
    public int minAddToMakeValid(String s) {
        int openCountToBalance = 0;
        int closeCountToBalance = 0;

        for (char symbol : s.toCharArray()) {
            if ('(' == symbol) {
                ++closeCountToBalance;
            } else {
                --closeCountToBalance;
            }

            if (-1 == closeCountToBalance) {
                ++openCountToBalance;
                closeCountToBalance = 0;
            }
        }

        return openCountToBalance + closeCountToBalance;
    }
}
